File: 第一章/learning_service/src/turtle_vel_cmd.cpp
#include <ros/ros.h>
#include <geometry_msgs/Twist.h>
#include <std_srvs/Trigger.h>

bool pubflag = false;

// service回调函数，输入参数req，输出参数res
bool commandCallback(std_srvs::Trigger::Request  &req,
         			std_srvs::Trigger::Response &res)
{
	pubflag = !pubflag;

    // 显示请求数据
    ROS_INFO("Publish turtle velocity command: %s", pubflag==true?"Yes":"No");

	// 设置反馈数据
	res.success = true;
	res.message = "Change turtle pubflag value successfully";

    return true;
}

int main(int argc, char **argv)
{
    // ROS节点初始化
    ros::init(argc, argv, "turtle_vel_cmd");

    // 创建节点句柄
    ros::NodeHandle nh;

    // 创建一个名为/turtle_command的server，注册回调函数commandCallback
    ros::ServiceServer cmd_service = nh.advertiseService("/turtle_cmd", commandCallback);

	// 创建一个Publisher，发布名为/turtle1/cmd_vel的topic，消息类型为geometry_msgs::Twist，队列长度10
	ros::Publisher turtle_vel_pub = nh.advertise<geometry_msgs::Twist>("/turtle1/cmd_vel", 1000);

    // 循环等待回调函数
    ROS_INFO("Ready to receive turtle velocity command.");

	geometry_msgs::Twist vel_msg;
	vel_msg.linear.x = 1.0;
	vel_msg.angular.z = 1.0;
	
	// 设置循环的频率
	ros::Rate loop_rate(10);

	while(ros::ok())
	{
		// 查看一次回调函数队列
    	ros::spinOnce();
		
		// 如果标志为true，则发布速度指令
		if(pubflag)
		{
			turtle_vel_pub.publish(vel_msg);
		}

		//按照循环频率延时
	    loop_rate.sleep();
	}

    return 0;
}

File: chapter 2/learning_service/scripts/turtle_vel_cmd.py
#!/usr/bin/env python
# coding:utf-8

import rospy
import threading
from geometry_msgs.msg import Twist
from std_srvs.srv import Trigger, TriggerResponse

pubflag = False;
turtle_vel_pub = rospy.Publisher('/turtle1/cmd_vel', Twist, queue_size=1000)

def commandCallback(req):
	global pubflag
	pubflag = bool(1-pubflag)

	rospy.loginfo("Publish turtle velocity command: %s", pubflag)
	
	return TriggerResponse(1, "Change turtle command state successfully")

def turtle_command_server():

    rospy.init_node('turtle_vel_cmd', anonymous=True)

    s = rospy.Service('/turtle_cmd', Trigger, commandCallback)
    
    rate = rospy.Rate(10) 

    vel_msg = Twist()
    vel_msg.linear.x = 1.0;
    vel_msg.angular.z = 1.0;

    print( "Ready to receive turtle command.")
    while 1:
        if pubflag:
            turtle_vel_pub.publish(vel_msg)
            print("The number of threads are: %d" %len(threading.enumerate()))
            rate.sleep()

    rospy.spin()

if __name__ == "__main__":
    turtle_command_server()